examples/hitl/rearrange_v2/rearrange_v2.py

```diff
@@ -306,31 +284,62 @@ def __init__(self, app_service: AppService):
                     gui_agent_controller=self._gui_agent_controllers[
                         user_index
                     ],
-                    client_helper=self._client_helper,
                     server_sps_tracker=self._sps_tracker,
                 )
             )
 
+        # Reset the environment immediately.
+        self.on_environment_reset(None)
+
+    def get_next_state(self) -> Optional[AppStateBase]:
+        # If cancelled, skip upload and clean-up.
+        if self._cancel or self._is_episode_finished():
+            return create_app_state_reset(self._app_service, self._app_data)
+        else:
+            return None
+
+    def on_enter(self):
+        super().on_enter()
+
+        user_index_to_agent_index_map: Dict[int, int] = {}
+        for user_index in range(len(self._user_data)):
+            user_index_to_agent_index_map[user_index] = self._user_data[
+                user_index
+            ].gui_agent_controller._agent_idx
+
+    def on_exit(self):
+        super().on_exit()
+
+    def _is_episode_finished(self) -> bool:
+        """
+        Determines whether all users have finished their tasks.
+        """
+        return all(
+            self._user_data[user_index].episode_finished
+            for user_index in self._users.indices(Mask.ALL)
+        )
```

Comment:
Interested to see how can I expose user-data for llm-agent to you?

Reply:
Probably not ideal, but the easiest solution would be to query the agent controller types at initialization and handle things from there.